File: src/cxx/VulkanLib/Device/Buffer/UniformBuffer.cpp
//
// Created by kgaft on 3/11/24.
//
#include "UniformBuffer.hpp"

UniformBuffer::UniformBuffer(std::shared_ptr<LogicalDevice> device, size_t bufferSize) {


    vk::BufferCreateInfo *createInfo = createInfos.getObjectInstance();
    createInfo->sType = vk::StructureType::eBufferCreateInfo;
    createInfo->usage = vk::BufferUsageFlagBits::eUniformBuffer;
    createInfo->size = bufferSize;

    createInfo->sharingMode = vk::SharingMode::eExclusive;
    base = new Buffer(device, createInfo,
                      vk::MemoryPropertyFlagBits::eHostVisible | vk::MemoryPropertyFlagBits::eHostCoherent);

    base->map(&mapPoint, 0, vk::MemoryMapFlags());
    createInfos.releaseObjectInstance(createInfo);
}

void *UniformBuffer::getMapPoint() {
    return mapPoint;
}

vk::Buffer UniformBuffer::getBuffer() {
    return base->getBuffer();
}

void UniformBuffer::destroy() {
    base->unMap();
    base->destroy();
    destroyed = true;
}

File: src/main/java/com/kgaft/VulkanLib/Device/Buffer/UniformBuffer.java
package com.kgaft.VulkanLib.Device.Buffer;

import com.kgaft.VulkanLib.Device.LogicalDevice.LogicalDevice;
import com.kgaft.VulkanLib.Utils.VkErrorException;
import org.lwjgl.PointerBuffer;
import org.lwjgl.vulkan.VkBufferCreateInfo;

import java.lang.instrument.IllegalClassFormatException;

import static org.lwjgl.vulkan.VK13.*;

public class UniformBuffer extends Buffer{
    public UniformBuffer(LogicalDevice device, long bufferSize) throws IllegalClassFormatException, VkErrorException {
        VkBufferCreateInfo createInfo = bufferCreateInfos.acquireObject();
        createInfo.sType$Default();
        createInfo.usage(VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT);
        createInfo.size(bufferSize);
        createInfo.sharingMode(VK_SHARING_MODE_EXCLUSIVE);
        super(device, createInfo, VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
        bufferCreateInfos.releaseObjectInstance(createInfo);
        map(mapPoint, 0, 0);
    }
    private PointerBuffer mapPoint = PointerBuffer.allocateDirect(1);

    public PointerBuffer getMapPoint() {
        return mapPoint;
    }

    @Override
    public void destroy() {
        unMap();
        super.destroy();
    }
}
